File: src/pocketcloud/network/packet/impl/normal/PlayerSwitchServerPacket.php
<?php

namespace pocketcloud\network\packet\impl\normal;

use pocketcloud\event\impl\player\PlayerSwitchServerEvent;
use pocketcloud\network\client\ServerClient;
use pocketcloud\network\Network;
use pocketcloud\network\packet\CloudPacket;
use pocketcloud\network\packet\utils\PacketData;
use pocketcloud\player\CloudPlayerManager;
use pocketcloud\server\CloudServerManager;
use pocketcloud\util\CloudLogger;

class PlayerSwitchServerPacket extends CloudPacket {

    public function __construct(private string $playerName = "", private string $newServer = "") {}

    public function encodePayload(PacketData $packetData): void {
        $packetData->write($this->playerName);
        $packetData->write($this->newServer);
    }

    public function decodePayload(PacketData $packetData): void {
        $this->playerName = $packetData->readString();
        $this->newServer = $packetData->readString();
    }

    public function getPlayerName(): string {
        return $this->playerName;
    }

    public function getNewServer(): string {
        return $this->newServer;
    }

    public function handle(ServerClient $client) {
        if (($player = CloudPlayerManager::getInstance()->getPlayerByName($this->playerName)) !== null) {
            if (($server = CloudServerManager::getInstance()->getServerByName($this->newServer)) !== null) {
                Network::getInstance()->broadcastPacket($this);
                CloudLogger::get()->debug("Player %s performed a server switch (%s -> %s)", false, $player->getName(), ($player->getCurrentServer()?->getName() ?? "NULL"), ($server?->getName() ?? "NULL"));
                (new PlayerSwitchServerEvent($player, $player->getCurrentServer(), $server));
                $player->setCurrentServer($server);
            }
        }
    }
}

File: cloud-network/src/main/java/de/pocketcloud/network/packet/impl/PlayerSwitchServerPacket.java
package de.pocketcloud.network.packet.impl;

import de.pocketcloud.api.network.packet.AuthenticatedPacket;
import de.pocketcloud.api.network.packet.CloudboundPacket;
import de.pocketcloud.api.network.packet.data.IPacketData;
import de.pocketcloud.network.packet.CloudPacket;
import lombok.Getter;
import lombok.NoArgsConstructor;

@NoArgsConstructor
@Getter
public final class PlayerSwitchServerPacket extends CloudPacket implements CloudboundPacket, AuthenticatedPacket {

    private String player;
    private String newServer;

    public PlayerSwitchServerPacket(String player, String newServer) {
        this.player = player != null ? player : "";
        this.newServer = newServer != null ? newServer : "";
    }

    @Override
    public void encodePayload(IPacketData packetData) {
        packetData.writeAll(player, newServer);
    }

    @Override
    public void decodePayload(IPacketData packetData) {
        this.player = packetData.readString();
        this.newServer = packetData.readString();
    }

    public static PlayerSwitchServerPacket create(String player, String newServer) {
        return new PlayerSwitchServerPacket(player, newServer);
    }
}
